test/lifecycle_test.dart

```diff
@@ -309,6 +309,50 @@ void main() {
       ]));
     });
 
+    test('updates state with correct lifecycle calls when `redraw` is called', () {
+      const Map initialState = const {
+        'initialState': 'initial',
+      };
+
+      final Map initialProps = unmodifiableMap({
+        'getInitialState': (_) => initialState
+      });
+
+      final Map newContext = const {};
+
+      final Map expectedProps = unmodifiableMap(
+          defaultProps,
+          initialProps,
+          emptyChildrenProps
+      );
+
+      _LifecycleTest component = getDartComponent(render(LifecycleTest(initialProps)));
+
+      component.lifecycleCalls.clear();
+
+      component.redraw();
+
+      expect(component.lifecycleCalls, equals([
+        matchCall('shouldComponentUpdateWithContext', args: [expectedProps, initialState, newContext],  state: initialState),
+        matchCall('componentWillUpdate',              args: [expectedProps, initialState],              state: initialState),
+        matchCall('componentWillUpdateWithContext',   args: [expectedProps, initialState, newContext],  state: initialState),
+        matchCall('render',                                                                             state: initialState),
+        matchCall('componentDidUpdate',               args: [expectedProps, initialState],              state: initialState),
+      ]));
+    });
+
+    test('properly handles a call to setState within another setState callback', () {
+      _LifecycleTest component = getDartComponent(render(LifecycleTest({})));
+
+      expect(() {
+        component.setState({'foo': 'bar'}, () {
+          component.lifecycleCalls.clear();
+
+          component.setState({'baz': 'foobar'}, () {});
```

Comment:
I know this test was mentioned earlier, but now I'm a little worried that this isn't fully testing what could go wrong with the state transition and the callbacks... (for example, a callback added inside a setState callback getting called before that transition occurs)

Something like this would provide more coverage around the inner callback being called, and the state of `state` in response to each call:
```dart
var expectedState1 = {}..addAll(initialState)..addAll({'foo', 'bar'};
var expectedState2 = {}..addAll(expectedState2)..addAll({'baz': 'foobar'});

component.setState({'foo': 'bar'}, expectAsync0(() {
  expect(component.state, expectedState1);
  component.setState({'baz': 'foobar'}, expectAsync0(() {
    expect(component.state, expectedState2);
  }));
}));

expect(component.lifecycleCalls, containsAllInOrder([
  matchCall('componentWillUpdate', args: [expectedProps, expectedState1]),
  matchCall('componentWillUpdate', args: [expectedProps, expectedState2]),
]));
```
Thoughts?